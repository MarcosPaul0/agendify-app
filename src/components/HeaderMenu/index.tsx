import { APP_ROUTES } from '@constants/appRoutes.constant';
import { COLORS } from '@constants/colors.constant';
import { useAuthContext } from '@contexts/AuthContext';
import { Link } from 'expo-router';
import {
  Clock,
  Briefcase,
  User,
  List,
  SignOut,
  Calendar,
} from 'phosphor-react-native';
import { useState } from 'react';
import {
  GestureResponderEvent,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import { Menu } from 'react-native-paper';
import { IHeaderMenuProps } from './interfaces/headerMenuProps.interface';

export function HeaderMenu({ businessId }: IHeaderMenuProps) {
  const { user, isAuthenticated, logout } = useAuthContext();

  const [menuIsVisible, setMenuIsVisible] = useState(false);
  const [menuAnchor, setMenuAnchor] = useState({ x: 0, y: 0 });

  function openMenu() {
    setMenuIsVisible(true);
  }

  function closeMenu() {
    setMenuIsVisible(false);
  }

  function handleMenuPress(event: GestureResponderEvent) {
    const { nativeEvent } = event;
    const anchor = {
      x: nativeEvent.pageX,
      y: nativeEvent.pageY,
    };

    setMenuAnchor(anchor);
    openMenu();
  }

  return (
    <View className="px-2 flex-row mx-2">
      <TouchableOpacity onPress={handleMenuPress}>
        <List size={24} />
      </TouchableOpacity>

      <Menu
        anchor={menuAnchor}
        visible={menuIsVisible}
        onDismiss={closeMenu}
        style={{ width: 160 }}
      >
        <View className="gap-2">
          <Link href={`${APP_ROUTES.MANAGER_USER}/${user?.id}`}>
            <View className="flex-row items-center gap-2">
              <User size={24} color={COLORS.BLUE_900} />
              <Text>Meu perfil</Text>
            </View>
          </Link>

          {businessId && (
            <Link
              className="flex-row items-center"
              href={`${APP_ROUTES.MANAGE_MY_BUSINESS}/${businessId}`}
            >
              <View className="flex-row items-center gap-2 ml-2">
                <Briefcase size={24} color={COLORS.BLUE_900} />
                <Text>Dados do neg√≥cio</Text>
              </View>
            </Link>
          )}

          {businessId && (
            <Link
              className="flex-row items-center"
              href={`${APP_ROUTES.MY_AVAILABILITY}/${businessId}`}
            >
              <View className="flex-row items-center gap-2 ml-2">
                <Clock size={24} color={COLORS.BLUE_900} />
                <Text>Disponibilidade</Text>
              </View>
            </Link>
          )}

          <Link
            className="flex-row items-center"
            href={APP_ROUTES.MY_APPOINTMENTS}
          >
            <View className="flex-row items-center gap-2 ml-2">
              <Calendar size={24} color={COLORS.BLUE_900} />
              <Text>Agendamentos</Text>
            </View>
          </Link>

          {isAuthenticated && (
            <TouchableOpacity
              className="flex-row items-center gap-2"
              onPress={logout}
            >
              <SignOut size={24} color={COLORS.BLUE_900} />
              <Text>Sair</Text>
            </TouchableOpacity>
          )}
        </View>
      </Menu>
    </View>
  );
}
