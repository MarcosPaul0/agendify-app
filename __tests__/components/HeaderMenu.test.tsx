import { render } from '@testing-library/react-native';
import { HeaderMenu } from '@components/HeaderMenu';

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn().mockImplementation((key: string) => `fake${key}token`),
  removeItem: jest.fn().mockImplementation((key: string) => `fake${key}token`),
  setItem: jest
    .fn()
    .mockImplementation(
      (key: string, value: string) => `fake${key}token${value}`
    ),
}));

describe('Header Menu', () => {
  it('Match with snapshot', async () => {
    const tree = render(<HeaderMenu businessId="fake-business-id" />);

    expect(tree).toMatchSnapshot();
  });
});
