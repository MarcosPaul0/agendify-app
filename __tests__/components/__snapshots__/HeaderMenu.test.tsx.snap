// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Menu Match with snapshot 1`] = `
<View
  style={
    [
      {
        "paddingLeft": 8,
        "paddingRight": 8,
      },
      {
        "flexDirection": "row",
      },
      {
        "marginLeft": 8,
        "marginRight": 8,
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="24"
      bbWidth="24"
      color="#000"
      focusable={false}
      height={24}
      meetOrSlice={0}
      minX={0}
      minY={0}
      size={24}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          [
            undefined,
            undefined,
            {},
          ],
          {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      tintColor="#000"
      vbHeight={256}
      vbWidth={256}
      width={24}
    >
      <RNSVGGroup>
        <RNSVGRect
          fill={null}
          height="256"
          propList={
            [
              "fill",
            ]
          }
          width="256"
          x="0"
          y="0"
        />
        <RNSVGLine
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="16"
          x1="40"
          x2="216"
          y1="128"
          y2="128"
        />
        <RNSVGLine
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="16"
          x1="40"
          x2="216"
          y1="64"
          y2="64"
        />
        <RNSVGLine
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="16"
          x1="40"
          x2="216"
          y1="192"
          y2="192"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <View
    collapsable={false}
  />
</View>
`;
